Docs: Add Terraform guide for Access Lists (#57938)
gravitational/teleport: docs/pages/identity-governance/access-lists/terraform.mdx

---
title: Managing Access Lists with Terraform
description: Learn how to manage Access Lists and their members with Terraform.
labels:
 - how-to
 - identity-governance
---

Is is possible to manage Access List and their members using Terraform. This guide will help you:

- Understand how to manage Access Lists with Terraform
- Familiarize with basic Access List types
- Understand how to manage Access List members, and which Access List types allow that and why

## How it works

The goal here is to create a simple organization structure represented with Access Lists using
Terraform. It will consist of:

- Access List representing a team
- Access List representing access to a resource (staging database)
- assigning the team access to the resource
- (optionally) assigning Okta group access to the resource

Access List management with Terraform is mostly done with 2 Terraform resources
`teleport_access_list` and `teleport_access_list_member`.

There are some nuances to members management. This is where concept of *Access List type* comes
into play. There are 2 key Access List types *default* and *static*.

> **default**

Reviewer: I would use a heading here since this is breaking up the document into subsections

  default Access List members with Terraform.

This makes sense, because during the audit process members can be removed from the list and there
is no way to reflect that in the Terraform manifests. So it doesn't make sense to manage members of
such lists in Terraform, because the source truth for the membership is the Teleport itself.

**static**

Access List of type static, have `.spec.type` set to "static". They differ from the default Access Lists:

- They don't support audit. `.spec.audit` field can be set, but it's completely ignored by Teleport.
- Their members can be managed by Terraform.

For static Access Lists, the source of truth for the membership is external (provisioned by
Terraform) so audit is not supported. The members should be reviewed at source, which are the
Terraform data sources or manifests and this process has to be external to Teleport.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx edition="Teleport Enterprise (v18.2.0 or higher)"!)
- Teleport [Terraform provider set up with credentials](../../zero-trust-access/infrastructure-as-code/terraform-provider/terraform-provider.mdx)

> ## Step 1/5. Create Access List for a team

Reviewer: ```suggestion
## Step 1/5. Create an Access List for a team
```
Grammar (article use)

```hcl
resource "teleport_access_list" "developers" {
  depends_on = [ teleport_role.noop ] # roles can't be deleted if they are in use
  header = {
    version = "v1"
    metadata = {
      name = "developers"
    }
  }
  spec = {
    type = "static" # type must be set to "static" to manage members with Terraform
    title       = "Developers"
    description = "Dev team."
    owners = [
      { name = "admin" },
    ]
    grants = {
      roles = ["noop"]
    }
  }
}

resource "teleport_access_list_member" "developers_alice" {
  header = {
    version = "v1"
    metadata = {
      name = "alice" # Teleport user name
    }
  }
  spec = {
    access_list     = teleport_access_list.developers.id # assign to "Developers" Access List
    membership_kind = 1 # 1 for "MEMBERSHIP_KIND_USER", 2 for "MEMBERSHIP_KIND_LIST"
  }
}

resource "teleport_access_list_member" "developers_bob" {
  header = {
    version = "v1"
    metadata = {
      name = "bob"
    }
  }
  spec = {
    access_list     = teleport_access_list.developers.id
    membership_kind = 1 # 1 for "MEMBERSHIP_KIND_USER", 2 for "MEMBERSHIP_KIND_LIST"
  }
}
```

> A few things to note here:

Author: I think it's ok to skim it. The first bullet point is the longest one, but it's just an introduction to the follow up steps. As for the rest I don't want to go into details how we convert Terraform data into proto structures. That's way out of scope.

This step will show how Access Lists of the default type can be employed in the structure. As a
quick reminder, the default Access Lists are supposed to be managed with the web UI and they
require periodic reviews performed by the list owners. The default Access List members cannot be
manged with Terraform, but the lists themselves can be created/modified and imported to Terraform.

Let's create a default Access List representing another team "DB Administrators" reusing the "noop"
role created in the previous steps:

```hcl
resource "teleport_access_list" "dbas" {
  depends_on = [ teleport_role.noop ] # roles can't be deleted if they are in use
  header = {
    version = "v1"
    metadata = {
      name = "dbas"
    }
  }
  spec = {
    # type is skipped, this is a regular Access Lists
    title       = "DBAs"
    description = "DB Administrators team."
    owners = [
      { name = "admin" },
    ]
    grants = {
      roles = ["noop"]
    }
    # audit settings can be skipped, the values below are the defaults
    audit = {
      recurrence = {
        frequency    = 6 # every 6 months
        day_of_month = 1 # first day of the month
      }
    }
  }
}
```

> The new "DBAs" team members can be managed using web UI. Teleport user "admin", as the Access List

Reviewer: We should use `code style` instead of double quotes for user names to make it clear that these have been specified in a configuration somewhere in this guide.

---
title: Managing Access Lists with Terraform
description: Learn how to manage Access Lists and their members with Terraform.
labels:
 - how-to
 - identity-governance
---

> Is is possible to manage Access List and their members using Terraform. This guide will help you:

Reviewer 2: ```suggestion
You can manage Access List and their members using Terraform. This guide will help you:
```

- Understand how to manage Access Lists with Terraform
- Familiarize with basic Access List types
- Understand how to manage Access List members, and which Access List types allow that and why

## How it works

> The goal here is to create a simple organization structure represented with Access Lists using

Reviewer 2: ```suggestion
We will create a simple organization structure represented with Access Lists using
```

- assigning the team access to the resource
- (optionally) assigning Okta group access to the resource

Access List management with Terraform is mostly done with 2 Terraform resources:
`teleport_access_list` and `teleport_access_list_member`.

There are some nuances to members management. This is where concept of *Access List type* comes
into play. There are 2 key Access List types: *default* and *static*.

### Access List of default type

The default Access List has the `.spec.type` field unset, or set to null or an empty string. Those are
regular Access Lists, like those created in the web UI. And as a consequence:

- They require auditing. Even though `.spec.audit` is not required, to be specified in the
  Terraform resource, the default value will be set and those lists have to be periodically
  reviewed in the web UI.
- Their members can be only managed with with the web UI and `tctl`. Is it not possible to managed
  default Access List members with Terraform.

This makes sense, because during the audit process members can be removed from the list and there
is no way to reflect that in the Terraform manifests. So it doesn't make sense to manage members of

> such lists in Terraform, because the source truth for the membership is the Teleport itself.

Author: I added this because that was a customer's concern that we don't support the reviews for static Access Lists. This tries to explain why. Let me know if you think it isn't needed regardless.

    }
  }
  spec = {
    access_list     = teleport_access_list.developers.id # assign to "Developers" Access List
    membership_kind = 1 # 1 for "MEMBERSHIP_KIND_USER", 2 for "MEMBERSHIP_KIND_LIST"
  }
}

resource "teleport_access_list_member" "developers_bob" {
  header = {
    version = "v1"
    metadata = {
      name = "bob"
    }
  }
  spec = {
    access_list     = teleport_access_list.developers.id
    membership_kind = 1 # 1 for "MEMBERSHIP_KIND_USER", 2 for "MEMBERSHIP_KIND_LIST"
  }
}
```

A few things to note here:

- The Developers list is just a container for the users which belong to a single team in the org.
  On its own it doesn't grant any permissions to its members, as it assigns only this "noop" role,
  but it is possible to nest this list in another Access List to inherit the permissions. That will
  be done in the follow up step.
- Each Access List member is managed as a separate resource in Terraform. This is how they are
  represented in Teleport backend. Teleport Terraform provider resources are very close
  representation of Terraform resources.
- As a consequence of the above `membership_kind` enum is a numeric type. `1` is for user members

>   and `2` is for nested Access List members. Both membership kinds will be showcased in this guide.

Author: Ok, the second one I wrote because there was a question in the discovery call why members are separate resources. But I agree it doesn't bring much value.

owner, will be automatically granted permissions to manage members of this list. `admin` will be
also required to review this list in the web UI every 6 months as defined in the *audit* section.
Because Access List members are separate resources in Teleport, modifying the members won't affect
the Teleport state for the `"teleport_access_list" "dbas"` resource.

Now, let's give "DBAs" the "Staging DBs access":

```hcl
resource "teleport_access_list_member" "db_access_staging_dbas" {
  header = {
    version = "v1"
    metadata = {
      name = teleport_access_list.dbas.id
    }
  }
  spec = {
    access_list     = teleport_access_list.db_access_staging.id
    membership_kind = 2 # 1 for "MEMBERSHIP_KIND_USER", 2 for "MEMBERSHIP_KIND_LIST"
  }
}
```

## Step 5/6. (optional) Give access to a Okta group

The Okta integration allows [synchronizing Okta groups and apps](../okta/app-and-group-sync.mdx) as
Teleport Access Lists. 

To give permissions to the Okta group represented as an Access List in Teleport, navigate to the Okta
Access List in the web UI, and from its URL (e.g.
`https://example.teleport.sh/web/accesslists/00gt3c8z9ukePm5uF697`) copy the last path segment. In

> this case `00gt3c8z9ukePm5uF697` - this is the name of the Okta Access List resource in Teleport.

Author: I double checked but I don't see any other way.

  }
}
```

## FAQ

### Can I import my existing Access List into Terraform and start managing their members with IaC?

This is usually not the case. Any Access List can be imported, but Access Lists created in the UI
are not static type, so their members can't be managed with Terraform. The same applies to Access
Lists created by an integration (e.g. Okta or Entra ID). An existing static list created by another
Terraform setup (with a different state) could be imported and then its members can be managed by
Terraform, but that's rather a rare use-case.

The recommended solution is to create a static Access List managed with Terraform and make this
list a member of the existing Access List. That way the members of the static Access List managed
with Terraform will inherit the grants of the existing Access List.

## Next Steps

- Learn how [nested Access Lists](./nested-access-lists.mdx) grants inheritance work.
- Familiarize yourself with
  [teleport_access_list](../../reference/terraform-provider/resources/access_list.mdx) and
  [teleport_access_list_member](../../reference/terraform-provider/resources/access_list_member.mdx)
  Terraform resources.

> 

Reviewer 2: Nit: Add links to Okta and Entra integrations.
